analyze panic after change partition prune mode from static to dynamic with analyze verison 1

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1. deploy a cluster with tikv 
```sql
set global tidb_analyze_version=1;
set tidb_analyze_version=1;
CREATE TABLE `t` (   `id` int(11) DEFAULT NULL,   `a` int(11) DEFAULT NULL, `b` int(11) DEFAULT NULL, `c` int(11) DEFAULT NULL ) PARTITION BY RANGE (`id`) (PARTITION `p0` VALUES LESS THAN (3),  PARTITION `p1` VALUES LESS THAN (7),  PARTITION `p2` VALUES LESS THAN (11));
insert into t values (1,1,1,1),(2,2,2,2),(4,4,4,4),(5,5,5,5),(6,6,6,6),(8,8,8,8),(9,9,9,9);
create index idxa on t (a);
create index idxb on t (b);
create index idxc on t (c);
analyze table t partition p0 index idxa;
analyze table t partition p1 index idxb;
analyze table t partition p2 index idxc;
set tidb_partition_prune_mode=dynamic;
analyze table t partition p0;
```
<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)
analyze successfully
### 3. What did you see instead (Required)
tidb panic
```bash
[2022/05/31 11:00:13.519 +08:00] [ERROR] [conn.go:1056] ["connection running loop panic"] [conn=2199023255955] [lastSQL="analyze table t partition p0"] [err="runtime error: invalid memory address or nil pointer dereference"] [stack="github.com/pingcap/tidb/server.(*clientConn).Run.func1\n\t/Users/aytrack/gitproject/pingcap/tidb/server/conn.go:1059\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:838\ngithub.com/pingcap/tidb/executor.(*ExecStmt).Exec.func1\n\t/Users/aytrack/gitproject/pingcap/tidb/executor/adapter.go:371\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:838\nruntime.panicmem\n\t/usr/local/go/src/runtime/panic.go:220\nruntime.sigpanic\n\t/usr/local/go/src/runtime/signal_unix.go:818\ngithub.com/pingcap/tidb/statistics.(*Index).TotalRowCount\n\t/Users/aytrack/gitproject/pingcap/tidb/statistics/histogram.go:1350\ngithub.com/pingcap/tidb/statistics.(*Table).GetStatsInfo\n\t/Users/aytrack/gitproject/pingcap/tidb/statistics/table.go:302\ngithub.com/pingcap/tidb/statistics/handle.(*Handle).mergePartitionStats2GlobalStats\n\t/Users/aytrack/gitproject/pingcap/tidb/statistics/handle/handle.go:475\ngithub.com/pingcap/tidb/statistics/handle.(*Handle).MergePartitionStats2GlobalStatsByTableID\n\t/Users/aytrack/gitproject/pingcap/tidb/statistics/handle/handle.go:405\ngithub.com/pingcap/tidb/executor.(*AnalyzeExec).Next\n\t/Users/aytrack/gitproject/pingcap/tidb/executor/analyze.go:210\ngithub.com/pingcap/tidb/executor.Next\n\t/Users/aytrack/gitproject/pingcap/tidb/executor/executor.go:319\ngithub.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelayExecutor\n\t/Users/aytrack/gitproject/pingcap/tidb/executor/adapter.go:665\ngithub.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelay\n\t/Users/aytrack/gitproject/pingcap/tidb/executor/adapter.go:516\ngithub.com/pingcap/tidb/executor.(*ExecStmt).Exec\n\t/Users/aytrack/gitproject/pingcap/tidb/executor/adapter.go:465\ngithub.com/pingcap/tidb/session.runStmt\n\t/Users/aytrack/gitproject/pingcap/tidb/session/session.go:2069\ngithub.com/pingcap/tidb/session.(*session).ExecuteStmt\n\t/Users/aytrack/gitproject/pingcap/tidb/session/session.go:1938\ngithub.com/pingcap/tidb/server.(*TiDBContext).ExecuteStmt\n\t/Users/aytrack/gitproject/pingcap/tidb/server/driver_tidb.go:230\ngithub.com/pingcap/tidb/server.(*clientConn).handleStmt\n\t/Users/aytrack/gitproject/pingcap/tidb/server/conn.go:2022\ngithub.com/pingcap/tidb/server.(*clientConn).handleQuery\n\t/Users/aytrack/gitproject/pingcap/tidb/server/conn.go:1876\ngithub.com/pingcap/tidb/server.(*clientConn).dispatch\n\t/Users/aytrack/gitproject/pingcap/tidb/server/conn.go:1371\ngithub.com/pingcap/tidb/server.(*clientConn).Run\n\t/Users/aytrack/gitproject/pingcap/tidb/server/conn.go:1121\ngithub.com/pingcap/tidb/server.(*Server).onConn\n\t/Users/aytrack/gitproject/pingcap/tidb/server/server.go:559"]
```
### 4. What is your TiDB version? (Required)
```sql
MySQL root@127.0.0.1:test> select tidb_version()\G
Reconnecting...
***************************[ 1. row ]***************************
tidb_version() | Release Version: v6.1.0-alpha-492-g15288cd55
Edition: Community
Git Commit Hash: 15288cd55861a572cf8a4c55f13d74f183eca4ac
Git Branch: release-6.1
UTC Build Time: 2022-05-31 02:58:44
GoVersion: go1.18
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```
<!-- Paste the output of SELECT tidb_version() -->

--- a/executor/analyze_test.go
+++ b/executor/analyze_test.go
@@ -3282,3 +3282,83 @@ PARTITION BY RANGE ( a ) (
 	require.Greater(t, tbl.Version, lastVersion)
 	require.Equal(t, 2, len(tbl.Indices[tableInfo.Indices[0].ID].Buckets))
 }
+
+func TestIssue35056(t *testing.T) {
+	store, dom, clean := testkit.CreateMockStoreAndDomain(t)
+	defer clean()
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("set @@session.tidb_analyze_version = 1")
+	createTable := `CREATE TABLE t (id int, a int, b varchar(10))
+PARTITION BY RANGE ( id ) (
+		PARTITION p0 VALUES LESS THAN (10),
+		PARTITION p1 VALUES LESS THAN (20)
+)`
+	tk.MustExec(createTable)
+	tk.MustExec("set @@session.tidb_partition_prune_mode = 'static'")
+	tk.MustExec("insert into t values (1,1,1),(2,2,2),(3,3,3),(4,4,4),(7,7,7),(9,9,9)")
+	tk.MustExec("insert into t values (11,11,11),(12,12,12),(14,14,14)")
+	h := dom.StatsHandle()
+	oriLease := h.Lease()
+	h.SetLease(1)
+	defer func() {
+		h.SetLease(oriLease)
+	}()
+	is := dom.InfoSchema()
+	h.HandleAutoAnalyze(is)
+	tk.MustExec("create index idxa on t (a)")
+	tk.MustExec("create index idxb on t (b)")
+	table, err := is.TableByName(model.NewCIStr("test"), model.NewCIStr("t"))
+	require.NoError(t, err)
+	tableInfo := table.Meta()
+	pi := tableInfo.GetPartitionInfo()
+	require.NotNil(t, pi)
+	tk.MustExec("analyze table t partition p0 index idxa")
+	tk.MustExec("analyze table t partition p1 index idxb")
+	tk.MustExec("set @@session.tidb_partition_prune_mode = 'dynamic'")
+	tk.MustExec("analyze table t partition p0")
+	tk.MustQuery("show warnings").Sort().Check(testkit.Rows(
+		"Warning 8244 Build table: `t` column: `id` global-level stats failed due to missing partition-level column stats, please run analyze table to refresh columns of all partitions",
+		"Warning 8244 Build table: `t` index: `idxa` global-level stats failed due to missing partition-level column stats, please run analyze table to refresh columns of all partitions",
+	))
+}
+
+func TestIssue35056Related(t *testing.T) {
+	store, dom, clean := testkit.CreateMockStoreAndDomain(t)
+	defer clean()
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("set @@session.tidb_analyze_version = 2")
+	createTable := `CREATE TABLE t (id int)
+PARTITION BY RANGE ( id ) (
+		PARTITION p0 VALUES LESS THAN (10),
+		PARTITION p1 VALUES LESS THAN (20)
+)`
+	tk.MustExec(createTable)
+	tk.MustExec("set @@session.tidb_partition_prune_mode = 'static'")
+	tk.MustExec("insert into t values (1),(2),(3),(4),(7),(9)")
+	tk.MustExec("insert into t values (11),(12),(14)")
+	h := dom.StatsHandle()
+	oriLease := h.Lease()
+	h.SetLease(1)
+	defer func() {
+		h.SetLease(oriLease)
+	}()
+	is := dom.InfoSchema()
+	h.HandleAutoAnalyze(is)
+	tk.MustExec("alter table t add column a int")
+	tk.MustExec("alter table t add column b int")
+	table, err := is.TableByName(model.NewCIStr("test"), model.NewCIStr("t"))
+	require.NoError(t, err)
+	tableInfo := table.Meta()
+	pi := tableInfo.GetPartitionInfo()
+	require.NotNil(t, pi)
+	tk.MustExec("analyze table t partition p0 columns id,a")
+	tk.MustExec("analyze table t partition p1 columns id,b")
+	tk.MustExec("set @@session.tidb_partition_prune_mode = 'dynamic'")
+	tk.MustExec("analyze table t partition p0")
+	tk.MustQuery("show warnings").Sort().Check(testkit.Rows(
+		"Note 1105 Analyze use auto adjusted sample rate 1.000000 for table test.t's partition p0",
+		"Warning 8244 Build table: `t` column: `a` global-level stats failed due to missing partition-level column stats, please run analyze table to refresh columns of all partitions",
+	))
+}

--- a/statistics/table.go
+++ b/statistics/table.go
@@ -298,11 +298,17 @@ func (t *Table) ColumnByName(colName string) *Column {
 // GetStatsInfo returns their statistics according to the ID of the column or index, including histogram, CMSketch, TopN and FMSketch.
 func (t *Table) GetStatsInfo(ID int64, isIndex bool) (int64, *Histogram, *CMSketch, *TopN, *FMSketch) {
 	if isIndex {
-		idxStatsInfo := t.Indices[ID]
-		return int64(idxStatsInfo.TotalRowCount()), idxStatsInfo.Histogram.Copy(), idxStatsInfo.CMSketch.Copy(), idxStatsInfo.TopN.Copy(), idxStatsInfo.FMSketch.Copy()
+		if idxStatsInfo, ok := t.Indices[ID]; ok {
+			return int64(idxStatsInfo.TotalRowCount()), idxStatsInfo.Histogram.Copy(), idxStatsInfo.CMSketch.Copy(), idxStatsInfo.TopN.Copy(), idxStatsInfo.FMSketch.Copy()
+		}
+		// newly added index which is not analyzed yet
+		return 0, nil, nil, nil, nil
 	}
-	colStatsInfo := t.Columns[ID]
-	return int64(colStatsInfo.TotalRowCount()), colStatsInfo.Histogram.Copy(), colStatsInfo.CMSketch.Copy(), colStatsInfo.TopN.Copy(), colStatsInfo.FMSketch.Copy()
+	if colStatsInfo, ok := t.Columns[ID]; ok {
+		return int64(colStatsInfo.TotalRowCount()), colStatsInfo.Histogram.Copy(), colStatsInfo.CMSketch.Copy(), colStatsInfo.TopN.Copy(), colStatsInfo.FMSketch.Copy()
+	}
+	// newly added column which is not analyzed yet
+	return 0, nil, nil, nil, nil
 }
 
 // GetColRowCount tries to get the row count of the a column if possible.

